core/src/sumeragi/mod.rs

```diff
@@ -53,42 +59,209 @@ trait Consensus {
+                incoming_message_receiver: Mutex::new(incoming_message_receiver),
+            },
+        })
+    }
+
+    pub fn update_metrics(&self, network: Addr<IrohaNetwork>) -> Result<()> {
+        use eyre::WrapErr;
+        use thiserror::Error;
+        let online_peers_count: u64 = self
+            .internal
+            .current_online_peers
+            .lock()
+            .unwrap()
+            .len()
+            .try_into()
+            .expect("casting usize to u64");
+
+        let mut wsv_guard = self.internal.wsv_for_public_use.lock().unwrap();
+
+        #[allow(clippy::cast_possible_truncation)]
+        if let Some(timestamp) = wsv_guard.genesis_timestamp() {
+            // this will overflow in 584942417years.
+            wsv_guard
+                .metrics
+                .uptime_since_genesis_ms
+                .set((current_time().as_millis() - timestamp) as u64)
+        };
+        let domains = wsv_guard.domains();
+        wsv_guard.metrics.domains.set(domains.len() as u64);
+        wsv_guard.metrics.connected_peers.set(online_peers_count);
+        for domain in domains {
+            wsv_guard
+                .metrics
+                .accounts
+                .get_metric_with_label_values(&[domain.id().name.as_ref()])
+                .wrap_err("Failed to compose domains")?
+                .set(domain.accounts().len() as u64);
+        }
+        Ok(())
+    }
+
+    pub fn latest_block_hash_for_use_by_block_sync(&self) -> HashOf<VersionedCommittedBlock> {
+        self.internal
+            .latest_block_hash_for_use_by_block_sync
+            .lock()
+            .unwrap()
+            .clone()
+    }
+
+    pub fn blocks_after_hash(
+        &self,
+        block_hash: HashOf<VersionedCommittedBlock>,
+    ) -> Vec<VersionedCommittedBlock> {
+        self.internal
+            .wsv_for_public_use
+            .lock()
+            .unwrap()
+            .blocks_after_hash(block_hash)
+    }
+
+    pub fn blocks_from_height(&self, block_height: usize) -> Vec<VersionedCommittedBlock> {
+        self.internal
+            .wsv_for_public_use
+            .lock()
+            .unwrap()
+            .blocks_from_height(block_height)
+    }
+
+    pub fn get_random_peer_for_block_sync(&self) -> Option<Peer> {
+        use rand::{prelude::SliceRandom, RngCore, SeedableRng};
+
+        let rng = &mut rand::rngs::StdRng::from_entropy();
+        let peers = self
+            .internal
+            .current_online_peers
+            .lock()
+            .expect("lock on online peers for get random peer")
+            .iter()
+            .map(|peer| Peer::new((*peer).clone()))
+            .collect::<Vec<Peer>>();
+        if peers.is_empty() {
+            return None;
+        } else {
+            let mut sorted_peers = peers;
+            sorted_peers.sort();
+            Some(sorted_peers[rng.next_u32() as usize % sorted_peers.len()].clone())
```

Comment:
I would suggest to sort `Vec<PeerId>` and only create `Peer` as last step this way we would remove redundant cloning.